pkg/driver/node.go

```diff
@@ -620,13 +620,41 @@ func (d *nodeService) nodePublishVolumeForFileSystem(req *csi.NodePublishVolumeR
 	mountOptions = collectMountOptions(fsType, mountOptions)
 
 	klog.V(4).Infof("NodePublishVolume: mounting %s at %s with option %s as fstype %s", source, target, mountOptions, fsType)
-	if err := d.mounter.Mount(source, target, fsType, mountOptions); err != nil {
-		if removeErr := os.Remove(target); removeErr != nil {
-			return status.Errorf(codes.Internal, "Could not remove mount target %q: %v", target, err)
+	/*
+		Checking if it's a mount point. There are three cases,
+		1. true, err when the directory does not exist or corrupted.
+		2. false, nil when the path is already mounted with a device.
+		3. true, nil when the path is not mounted with any device.
+	*/
+	notMnt, err := d.mounter.IsLikelyNotMountPoint(target)
+	if err != nil && !os.IsNotExist(err) {
+		//Checking if the path exists and error is related to Corrupted Mount, in that case, the system could unmount and mount.
+		_, pathErr := mountutils.PathExists(target)
+		if pathErr != nil && mountutils.IsCorruptedMnt(pathErr) {
+			klog.V(4).Infof("Target path %q is a corrupted directory", target)
```

Comment:
are we not returning an error if target path is a corrupted directory?

Reply:
Chances are that the device mounted or the mount point is corrupted because of improper unmounting. Hence in the next phase we try to read the directory and unmount it.